proposals/0152-precompiles.md

```diff
@@ -0,0 +1,270 @@
+    }
+    instruction_data = all_tx_data.instruction_datas[instruction_index]
+  }
+
+  start = offset
+  end = offset + length
+  if end > instruction_data_length {
+    return Error
+  }
+
+  return instruction_data[start..end]
+}
+```
+
+### Changes to `Ed25519SigVerify111111111111111111111111111`
+
+**Summary.**
+
+- Change #1. Replace sigverify function with Dalek `strict_verify()`,
+  the same used for transactions sigverify.
+
+- Change #2. Implement step 3 above: "If `num_signatures` is 0, return error."
+
+**Context.**
+
+In Solana, transactions use Ed25199 signatures, and are validated using
+the so called **strict verify**.
+Compared to "RFC verify", strict verify enforces extra checks against
+(certain types of) malleability.
+
+The Ed25199 precompile currently implements a non-strict verify, so with
+Change #1 we'll make it compatible with the way Solana verifies signatures.
+
+Moreover, the Ed25519 precompile accepts a payload of `[0, 0]` as valid
+(for no good reason), so Change #2 will prevent this anomaly.
+
+Finally, the Ed25519 precompile interleaves retrieving instruction data
+and parsing data types such as signatures and public keys.
+While this goes against this specification and creates unnecessary complexity
+in the return error code, we recommend to NOT change the internal behavior
+(as the return error code doesn't really matter).
+
+**FAQ.**
+
+- **Q: Why does the Ed25199 precompile currently use `verify` instead of
+  `strict_verify`?**
+  A: No good reason, it was built without noticing the difference.
+
+- **Q: If we switch to `strict_verify`, will some of the existing signatures
+  break verification?**
+  A: All signatures created by a "regular" library, i.e. following RFC,
+  pass both `verify` and `strict_verify`.
+  Only carefully crafted signatures can pass `verify` and not `strict_verify`.
+  So this won't break any honest use case.
+
+- **Q: Why not leaving it as is?**
+  A: The `verify` is not well specified. In fact, it's behavior is slightly
+  different in the older version of Dalek that Solana currently uses, versus
+  the latest version of the same library.
+  Trying to replicate all the edge cases is different validators is an
+  unnecessary effort, not worth the risk of exposing different behaviors.
+
+### Changes to `KeccakSecp256k11111111111111111111111111111`
+
+**Summary.**
+
+- Change #3. Implement step 3 above: "If `num_signatures` is 0, return error."
+
+**Context.**
+
+The KeccakSecp256k1 precompile currently accepts an input of `[0]`, as
+in "verify 0 signatures", which is a useless instruction.
+With Change #3 we'll avoid this anomaly.
+
+We note that the KeccakSecp256k1 precompile has a slightly different
+struct for offset, with instruction indexes of a single byte (and, as
+a result, no special value of `0xFFFF` to indicate the "current instruction").
+This is for historical reasons, and since modifying it would break
+existing users, we recommend to NOT change the existing behavior.
```

Comment:
This is not true. The struct is defined as
```rust
#[derive(Default, Serialize, Deserialize, Debug, Eq, PartialEq)]
pub struct SecpSignatureOffsets {
    /// Offset to 64-byte signature plus 1-byte recovery ID.
    pub signature_offset: u16,
    /// Within the transaction, the index of the instruction whose instruction data contains the signature.
    pub signature_instruction_index: u8,
    /// Offset to 20-byte Ethereum address.
    pub eth_address_offset: u16,
    /// Within the transaction, the index of the instruction whose instruction data contains the address.
    pub eth_address_instruction_index: u8,
    /// Offset to start of message data.
    pub message_data_offset: u16,
    /// Size of message data in bytes.
    pub message_data_size: u16,
    /// Within the transaction, the index of the instruction whose instruction data contains the message.
    pub message_instruction_index: u8,
}
```
The padding rules mean that a padding byte is introduced after each `u8`. So, by changing every `u8` to `u16` the layout does not change. Assuming that the clients insert zeros for the padding byte, nothing will break. 

So I think the offsets structs can be made identical.

I have not tested this.